airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py

```diff
@@ -838,14 +839,23 @@ def primary_keys_for_records(streams, records):
     for stream in streams_with_primary_key:
         stream_records = [r for r in records if r.stream == stream.stream.name]
         for stream_record in stream_records:
-            pk_values = {}
-            for pk_path in stream.stream.source_defined_primary_key:
-                pk_value = reduce(lambda data, key: data.get(key) if isinstance(data, dict) else None, pk_path, stream_record.data)
-                pk_values[tuple(pk_path)] = pk_value
-
+            pk_values = _extract_primary_key(stream_record.data, stream.stream.source_defined_primary_key)
             yield pk_values, stream_record
 
 
+def _extract_primary_keys(records: Iterable[Mapping[str, Any]], primary_key: List[str]) -> Iterable[Tuple[Any, Mapping[str, Any]]]:
```

Comment:
```suggestion
def _extract_primary_keys(records: Iterable[Mapping[str, Any]], primary_key: Tuple[str]) -> Iterable[Tuple[Any, Mapping[str, Any]]]:
```
Would it be better to use a tuple instead of list here?

Reply:
Actually, the typing comes for the airbyte_protocol.py and is defined as:

```
    source_defined_primary_key: Optional[List[List[str]]] = Field(
        None,
        description='If the source defines the primary key, paths to the fields that will be used as a primary key. If not provided by the source, the end user will have to specify the primary key themselves.',
    )
```

I'll update this to align with the `source_defined_primary_key` definition. The return type was also inadequate so I've also updated this. Thanks for pointing this out!